kubernetes/src/main/java/com/linecorp/armeria/client/kubernetes/endpoints/KubernetesEndpointGroup.java

```diff
@@ -0,0 +1,439 @@
+                    return;
+                }
+                logger.warn("{} service watcher is closed. (namespace: {})", namespace, serviceName, cause);
+            }
+        };
+
+        if (namespace == null) {
+            return client.services().withName(serviceName).watch(watcher);
+        } else {
+            return client.services().inNamespace(namespace).withName(serviceName).watch(watcher);
+        }
+    }
+
+    private Watch watchPod(Map<String, String> selector) {
+        final Watcher<Pod> watcher = new Watcher<Pod>() {
+            @Override
+            public void eventReceived(Action action, Pod resource) {
+                if (closed) {
+                    return;
+                }
+                if (action == Action.ERROR || action == Action.BOOKMARK) {
+                    return;
+                }
+                final String podName = resource.getMetadata().getName();
+                final String nodeName = resource.getSpec().getNodeName();
+                if (podName == null || nodeName == null) {
+                    logger.debug("Pod or node name is null. pod: {}, node: {}", podName, nodeName);
+                    return;
+                }
+
+                switch (action) {
+                    case ADDED:
+                    case MODIFIED:
+                        podToNode.put(podName, nodeName);
+                        break;
+                    case DELETED:
+                        podToNode.remove(podName);
+                        break;
+                    default:
+                }
+                maybeUpdateEndpoints();
+            }
+
+            @Override
+            public void onClose(WatcherException cause) {
+                if (closed) {
+                    return;
+                }
+
+                logger.warn("Pod watcher for {}/{} is closed.", namespace, serviceName, cause);
+            }
+        };
+
+        if (namespace == null) {
+            return client.pods().withLabels(selector).watch(watcher);
+        } else {
+            return client.pods().inNamespace(namespace).withLabels(selector).watch(watcher);
+        }
+    }
+
+    /**
+     * Fetches the internal IPs of the node.
+     */
+    private Watch watchNodes() {
+        final Watcher<Node> watcher = new Watcher<Node>() {
+            @Override
+            public void eventReceived(Action action, Node node) {
+                if (closed) {
+                    return;
+                }
+
+                if (action == Action.ERROR || action == Action.BOOKMARK) {
+                    return;
+                }
+
+                final String nodeName = node.getMetadata().getName();
+                switch (action) {
+                    case ADDED:
+                    case MODIFIED:
+                        final String nodeIp = node.getStatus().getAddresses().stream()
+                                                  .filter(address -> "InternalIP".equals(address.getType()))
```

Comment:
Question) Does `ExternalIP`/`HostName` not work? I think it's fine to only handle `InternaIP` for now, but just in case someone asks.